src/directory.rs

```diff
@@ -330,11 +341,11 @@ fn test_diff_display() {
     assert_eq!(
         diff_5.to_string(),
         [
-            "A bar",
             "D a/foo",
+            "A bar",
             "A baz/bar",
-            "D baz/foo",
             "M baz/baz",
+            "D baz/foo",
             "D foo",
             ""
         ]
```

Comment:
This is the only way I could think of controlling the ordering: now pushing all the paths into an intermediate map and then iterating through that and printing each item. The paths are in alphabetical order (which I think is close to how `git diff --name-status` works although I think that outputs files sorted by directory _then_ filename, maybe we'd want to do that too? It does seem useful to distinguish files from folders in printed output, but maybe that could be left to an "even prettier" diff PR 😄). Alternatively they could be sorted or grouped by added/modified/deleted status (is that what you meant here? https://github.com/SamuelSchlesinger/version-control/pull/12#discussion_r1191546662).

The additional `diff_paths` map feels a bit awkward but not sure what else to use, since I don't _think_ there's a sensible ordering of output that would result just by relying on traversing the diff and using `writeln` directly on each leaf.